src/globals.c

```diff
@@ -376,6 +379,147 @@ int size_var(var_t *var)
     return s;
 }
 
+/* linked list */
+fn_t *add_fn()
+{
+    fn_t *n = calloc(1, sizeof(fn_t));
+
+    if (!FUNC_LIST.head) {
+        FUNC_LIST.head = n;
+        FUNC_LIST.tail = n;
+        return n;
+    }
+    FUNC_LIST.tail->next = n;
+    FUNC_LIST.tail = n;
+    return n;
+}
+
+BB *create_BB(block_t *parent)
+{
+    BB *bb = calloc(1, sizeof(BB));
+
+    int i;
+    for (i = 0; i < MAX_BB_PRED; i++) {
+        bb->prev[i].bb = NULL;
+        bb->prev[i].type = NEXT;
+    }
+    bb->scope = parent;
+    bb->belong_to = parent->func->fn;
+    return bb;
+}
+
+void connect_BB(BB *pred, BB *succ, Block_Type_t type)
+{
+    if (!pred)
+        abort();
+    if (!succ)
+        abort();
+
+    int i = 0;
+    while (succ->prev[i].bb)
+        i++;
+
+    if (i > MAX_BB_PRED - 1) {
+        printf("Error: too many predecessors\n");
+        abort();
+    }
+
+    succ->prev[i].bb = pred;
+    succ->prev[i].type = type;
+
+    switch (type) {
+    case NEXT:
+        pred->next = succ;
+        break;
+    case THEN:
+        pred->then_ = succ;
+        break;
+    case ELSE:
+        pred->else_ = succ;
+        break;
+    default:
+        abort();
+    }
+}
+
+/* the pred-succ pair should only have one connection */
+void disconnect_BB(BB *pred, BB *succ)
+{
+    int i;
+    for (i = 0; i < MAX_BB_PRED; i++) {
+        if (succ->prev[i].bb == pred) {
+            switch (succ->prev[i].type) {
+            case NEXT:
+                pred->next = NULL;
+                break;
+            case THEN:
+                pred->then_ = NULL;
+                break;
+            case ELSE:
+                pred->else_ = NULL;
+                break;
+            default:
+                abort();
+            }
+
+            succ->prev[i].bb = NULL;
+            break;
+        }
+    }
+}
+
+void add_BB_symbol(BB *bb, var_t *var)
+{
+    symbol_t *sym;
+    for (sym = bb->symbol_list.head; sym; sym = sym->next)
+        if (sym->var == var)
+            return;
+
+    sym = calloc(1, sizeof(symbol_t));
+    sym->var = var;
+
+    if (!bb->symbol_list.head) {
+        sym->index = 0;
+        bb->symbol_list.head = sym;
+        bb->symbol_list.tail = sym;
+    } else {
+        sym->index = bb->symbol_list.tail->index + 1;
+        bb->symbol_list.tail->next = sym;
+        bb->symbol_list.tail = sym;
+    }
+}
+
+/* linked list */
+void add_inst(block_t *block,
+              BB *bb,
+              opcode_t op,
+              var_t *rd,
```

Comment:
Would it be better to reference a structure rather than passing these parameters?

Reply:
The old first phase IR will be and replaced by SSA CFG and dropped in the future.